feat(tactic/lift): add lift tactic

fpvandoorn opened the pull request:
TO CONTRIBUTORS:

Make sure you have:

  * [x] reviewed and applied the coding style: [coding](https://github.com/leanprover/mathlib/blob/master/docs/contribute/style.md), [naming](https://github.com/leanprover/mathlib/blob/master/docs/contribute/naming.md)
  * [x] reviewed and applied [the documentation requirements](https://github.com/leanprover/mathlib/blob/master/docs/contribute/doc.md)
  * [x] for tactics:
     * [x] added or adapted documentation in [tactics.md](https://github.com/leanprover/mathlib/blob/master/docs/tactics.md)
     * [x] write an example of use of the new feature in [tactics.lean](https://github.com/leanprover/mathlib/blob/master/tests/tactics.lean)
  * [x] make sure definitions and lemmas are put in the right files
  * [x] make sure definitions and lemmas are not redundant

[Zulip link](https://leanprover.zulipchat.com/#narrow/stream/113488-general/topic/tactics.20for.20numbers) 

For reviewers: [code review check list](https://github.com/leanprover/mathlib/blob/master/docs/contribute/code-review.md)

fpvandoorn committed:
start on lift_to tactic

fpvandoorn committed:
finish lift tactic

fpvandoorn committed:
add instance to lift rat to int

this required me to move some lemmas from rat/order to rat/basic which had nothing to do with the order on rat

fpvandoorn committed:
move test to test/tactic.lean

fpvandoorn committed:
add header and documentation

cipher1024 commented on src/tactic/lift.lean:
```diff
@@ -0,0 +1,114 @@
+
+/-- A class specifying that you can lift elements from `α` to `β` assuming `cond` is true.
+  Used by the tactic `lift`. -/
+class can_lift (α : Type u) (β : Type v) : Type (max u v) :=
```
You could declare this class as `class can_lift (α : Type u) (β : Type v) [has_coe α β] : Type (max u v) :=` so that you can use the standard coercion mechanism instead of `coe` declared in this class.

fpvandoorn replied:
What if we want to use the lift tactic in cases where we don't want a coercion from `β` to `α`? 
I think the lift tactic could be more general. 

For example: you add an instance of `can_lift (α ⊕ β) α` if you have a proof that `∀y, (x : α ⊕ β) ≠ sum.inr y`. Doing this with `lift` is probably not really easier than just doing cases on `x`, but it is possible. And there are probably situations where it is actually useful. Maybe you have a type that is equivalent to `α ⊕ β` (or maybe you just have a map from `α ⊕ β`), where a `can_lift` instance would be easier than going along the equivalence yourself.

This tactic might also occasionally be useful when you have an equivalence `α ≃ β` and want to move elements between `α` and `β`.

cipher1024 replied:
That's fair. I did not consider those use cases. In your examples, `can_lift.coe` seems to disappear` after using `lift`. Does your tactic make sure of that?

fpvandoorn replied:
Yes, I run `dsimp only with can_lift` on all relevant goals/hypotheses (and even in some error messages) - search the tactic for `dsimp`. That's why the `can_lift` instance has the `can_lift` attribute.
```
@[can_lift] instance : can_lift ℤ ℕ
```

cipher1024 replied:
Why not just by default include every `can_lift` instance in that `simp` set?

fpvandoorn replied:
how do I do that?

cipher1024 replied:
`simp` is a bit tricky to script. What I would do is use `simp_target` or `simp_hyp` (depending on what you're trying to rewrite) and use `mk_simp_set` to construct the `simp_lemmas` argument. When you call `mk_simp_set`, you give it a list of `simp_arg_type` built from the `pexpr` of all the instances of `can_lift`

fpvandoorn replied:
And I get all these instances by using ```attribute.get_instances `instance```, and for all of them looking at the target of the type? 
Is that efficient enough? That seems like it could be a performance hit for the very minor convenience of not having to write `@[can_lift]`. I'm leaning towards just using `@[can_lift]`.

cipher1024 replied:
You can index those instances with this kind of code:

```lean
open tactic

@[user_attribute]
meta def foo : user_attribute (list name) := 
{ name := "a",
  descr := "d",
  cache_cfg := { mk_cache := λ _,
    do{ ls ← attribute.get_instances `instance,
        ls.mfilter $ λ l, 
        do { (_,t) ← mk_const l >>= infer_type >>= mk_local_pis, 
         return $ t.is_app_of `decidable_eq } },
  dependencies := [`instance] } }

-- @[a]
instance decidable_eq.foo : decidable_eq ℕ := sorry

run_cmd do 
l ← foo.get_cache,
trace l.length
```

cipher1024 replied:
That code nominally declares an attribute that you never have to use. The attribute only serves as a location for an instance cache. The cache is rebuilt every time you declare an instance and the cache_cfg code keeps only `decidable_eq` instances (or `can_lift` in your case).

fpvandoorn replied:
Ah, that is pretty nice. I didn't think of using the attribute cache for this. Thank you for writing the `cache_cfg` for this attribute.

I now used this code (and added another test checking that it works properly).

robertylewis replied:
I agree, this is clever!

cipher1024 replied:
I'm pretty happy with it myself. I only recently managed to understand what `dependencies` does.

robertylewis commented on docs/tactics.md:
```diff
@@ -1053,3 +1053,29 @@ localized "attribute [simp] le_refl" in le
 
 `rotate` moves the first goal to the back. `rotate n` will do this `n` times.
+
+### lift
```
Could you illustrate this with some examples, printing the tactic state before and after calls? I get the idea, but it's a little hard to follow with `h t p type_p k hk` floating around.

fpvandoorn replied:
Yes, it wasn't written very well. I rewrote it from scratch using only examples which get increasingly complicated. Is this better, or do you want even more concrete examples?

robertylewis replied:
Thanks, this looks much nicer! You could copy the test here, and rewrite the guards as comments to show what's going on. I think it would be helpful but not completely necessary.

fpvandoorn replied:
Added some examples.

robertylewis commented on src/tactic/lift.lean:
```diff
@@ -0,0 +1,114 @@
+
+namespace tactic
+
+meta def lift (p : pexpr) (t : pexpr) (h : option pexpr) (n : list name) : tactic unit :=
```
I find the behavior of this tactic a little hard to follow. Are there reasonable subroutines that you could factor out? Maybe lines 45-55, for instance? I can imagine this being tricky to re-parse if we need to update it down the line. At least a few more comments would help.

fpvandoorn replied:
Better now?

robertylewis replied:
Much cleaner, thanks! `lift` needs its own doc string. (Since the main interface is the interactive version, the doc string should primarily just say what the arguments are.)

fpvandoorn replied:
done

robertylewis commented on src/tactic/lift.lean:
```diff
@@ -0,0 +1,114 @@
+    else get_unused_name `h,
+  temp_nm ← get_unused_name,
+  temp_e ← note temp_nm none prf_ex,
+  dsimp (interactive.loc.ns [temp_nm]),
```
E.g. this `dsimp` could use a comment. I wasn't sure what was happening here until I saw your answer to Simon.

fpvandoorn committed:
add more/better documentation

fpvandoorn committed:
typo

robertylewis commented on src/tactic/lift.lean:
```diff
@@ -0,0 +1,128 @@
+    (but first call dsimp on the expression in the message).
+  If the proof was not specified, we create assert it as a local constant.
+  (The name of this local constant doesn't matter, since `lift` will remove it from the context) -/
+meta def get_lift_prf (h : option pexpr) (old_tp new_tp inst e : expr) : tactic expr :=
```
The doc string should explicitly say what each of these arguments are. Between the names and description, I can mostly figure it out, but it takes some thought.

fpvandoorn replied:
done

fpvandoorn committed:
more documentation

fpvandoorn committed:
rewrite, minor

fpvandoorn committed:
move import

fpvandoorn committed:
remove can_lift attribute

now we automatically construct the simp set used to simplify
Thanks to @cipher1024 for the idea and writing the main part of this code

robertylewis reviewed (commented):
This looks great. One pedantic question before merging: I just tested it out and accidentally wrote a `has_lift` instance instead of `can_lift`. This is slightly confusing, but no big deal. My bigger worry is that they go in opposite directions -- the lift of `has_lift` moves a nat to an int. Is `has_lift` really going "up"? Is there a better name? I don't have a suggestion right now but it's late.

fpvandoorn committed:
remove occurrence of [can_lift]

mergify[bot] committed:
Merge branch 'master' into lift_to
